jenkinsci/jenkins
# [JENKINS-55582] Converting bundled modules to detached plugins

See [JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582). While the `jenkins-module` packaging remains an option for hypothetical use cases such as a custom `ConfidentialStore` implementation, it turns out that none of the _existing_ modules had any compelling reason to be modules: they were built and maintained like plugins with some `@Extension`s that could be loaded from a regular plugin class loader during the regular initialization sequence. Distributing them as modules offered no benefit, and had several drawbacks: complexity and inconsistency; difficulty updating functionality quickly (modules had to be delivered on the core LTS cycle); and a confusing mess in plugin POMs needing access to these classes, since you needed to declare a `provided` dependency on the exact version that happened to be bundled in your currently selected `jenkins.version`.

The existing “split plugin” system, for retaining (binary) compatibility of other plugins that might be depending on these classes, sufficed to mark the split point just like any classes moved out of core. The only real difference is that the handful of consuming plugins (e.g., callers of `InstanceIdentity`) were _explicitly_ including the dependency before. I have not changed `groupId`/`artifactId` so source updates on the caller side should just look like a plain update, except now you can switch to the regular `compile` scope (i.e., the default) and you will get an explicit entry in `Plugin-Dependencies` as you should.

The JNLP 4 protocol still needs the `instance-identity` plugin loaded in order to work (though in practice you would have a hard time disabling any of these new plugins in the near future, due to automatic dependencies). Resurrecting #2875 would clean this up. Other than that, there is very little in the way of code changes here; it is mostly rearranging some metadata.

Subsumes #3987 to aid testing. Downstream of

* https://github.com/jenkinsci/instance-identity-module/pull/17
* https://github.com/jenkinsci/ssh-cli-auth-module/pull/9
* https://github.com/jenkinsci/sshd-module/pull/29
* https://github.com/jenkinsci/slave-installer-module/pull/5
* https://github.com/jenkinsci/windows-slave-installer-module/pull/24
* https://github.com/jenkinsci/launchd-slave-installer-module/pull/3
* https://github.com/jenkinsci/upstart-slave-installer-module/pull/2
* https://github.com/jenkinsci/systemd-slave-installer-module/pull/3

### Manual testing

[This page](https://github.com/jenkinsci/systemd-slave-installer-module#testing) describes how to see one the newly detached agent installer plugins being used in a realistic class loader environment. If you first run, say, 2.164.2 and then upgrade to the WAR from this branch, the detached plugins are all installed automatically. The same is true if _any other_ plugin (built against Jenkins 2.173 or earlier) is installed.

SSH functionality was also tested. With the new `sshd` plugin enabled, you can configure a port, attach your public key to your user configuration, and run commands like

```bash
ssh -p 2222 admin@localhost who-am-i
```

You can also install `workflow-cps-global-lib` and use the (now deprecated) in-master Git server, `workflowLibs.git`, via SSH protocol.

### Proposed changelog entries

* Some components of Jenkins which were previously packaged as “modules” are now regular plugins, and can be managed normally including updating from the update center. These components handle an identity for inbound (“JNLP”) agent connections; an SSH server; and hooks to offer service installation for JNLP agents.
* “Detached” plugins—those split out of Jenkins core at some point, including the former “modules”—will now be installed upon Jenkins startup when needed as implied dependencies of other plugins which were already present. This simplifies compatibility for specialized installation scenarios not using the update center, such as when Jenkins is run from a Docker image prepackaged with some plugins.

### Desired reviewers

@jenkinsci/code-reviewers
